src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -0,0 +1,667 @@
+
+            if (null != Interlocked.CompareExchange(ref s_globalInstanceForTrackerSupport, instance, null))
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_ResetGlobalComWrappersInstance);
+            }
+        }
+
+        /// <summary>
+        /// Register a <see cref="ComWrappers" /> instance to be used as the global instance for marshalling in the runtime.
+        /// </summary>
+        /// <param name="instance">Instance to register</param>
+        /// <remarks>
+        /// This function can only be called a single time. Subsequent calls to this function will result
+        /// in a <see cref="System.InvalidOperationException"/> being thrown.
+        ///
+        /// Scenarios where this global instance may be used are:
+        ///  * Usage of COM-related Marshal APIs
+        ///  * P/Invokes with COM-related types
+        ///  * COM activation
+        /// </remarks>
+        public static void RegisterForMarshalling(ComWrappers instance)
+        {
+            if (instance == null)
+                throw new ArgumentNullException(nameof(instance));
+
+            if (null != Interlocked.CompareExchange(ref s_globalInstanceForMarshalling, instance, null))
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_ResetGlobalComWrappersInstance);
+            }
+        }
+
+        /// <summary>
+        /// Get the runtime provided IUnknown implementation.
+        /// </summary>
+        /// <param name="fpQueryInterface">Function pointer to QueryInterface.</param>
+        /// <param name="fpAddRef">Function pointer to AddRef.</param>
+        /// <param name="fpRelease">Function pointer to Release.</param>
+        protected internal static void GetIUnknownImpl(out IntPtr fpQueryInterface, out IntPtr fpAddRef, out IntPtr fpRelease)
+            => GetIUnknownImplInternal(out fpQueryInterface, out fpAddRef, out fpRelease);
+
+        internal static int CallICustomQueryInterface(object customQueryInterfaceMaybe, ref Guid iid, out IntPtr ppObject)
+        {
+            var customQueryInterface = customQueryInterfaceMaybe as ICustomQueryInterface;
+            if (customQueryInterface is null)
+            {
+                ppObject = IntPtr.Zero;
+                return -1; // See TryInvokeICustomQueryInterfaceResult
+            }
+
+            return (int)customQueryInterface.GetInterface(ref iid, out ppObject);
+        }
+
+        private static unsafe int ComputeThisPtrForDispatchSection(int dispatchCount)
+        {
+            // For 64 bit architeture that always would be the case.
+            const int EntriesPerThisPtr = 1;
+            return (dispatchCount / EntriesPerThisPtr) + ((dispatchCount % EntriesPerThisPtr) == 0 ? 0 : 1);
+        }
+
+        [UnmanagedCallersOnly]
+        internal static unsafe int ABI_QueryInterface(IntPtr ppObject, ref Guid guid, out IntPtr returnValue)
```

Comment:
```suggestion
        internal static unsafe int ABI_QueryInterface(IntPtr ppObject, Guid* guid, IntPtr* returnValue)
```
UnmanagedCallersOnly methods should only have unmanaged types in the signature. I am surprised that this actually compiles.